File: LeetCode/141_LinkedListCycle.cc
// https://leetcode.com/problems/linked-list-cycle/
// 2022 3.21

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        // two pointer
        ListNode *slow=head, *fast=head;
        while(fast and fast->next){
            fast=fast->next->next;
            slow=slow->next;
            if(slow==fast){
                return true;
            }
        }
        return false;
    }
};

File: LeetCode/141_LinkedListCycle.go
// https://leetcode.com/problems/linked-list-cycle/
// 2022年07月24日 22:29:01

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func hasCycle(head *ListNode) bool {
    // v1
    slow,fast:=head,head
    for fast!=nil && fast.Next!=nil{
        fast=fast.Next.Next
        slow=slow.Next
        if fast==slow{
            return true
        }
    }
    return false
}